programs/marinade-finance/src/instructions/user/withdraw_stake_account.rs

```diff
@@ -0,0 +1,302 @@
+        validator_index: u32,
+        msol_amount: u64,
+    ) -> Result<()> {
+        require!(!self.state.paused, MarinadeError::ProgramIsPaused);
+        require!(!self.state.withdraw_stake_account_enabled, MarinadeError::WithdrawStakeAccountIsNotEnabled);
+
+        // record  for event
+        let user_msol_balance = self.burn_msol_from.amount;
+        // save msol price source
+        let total_virtual_staked_lamports = self.state.total_virtual_staked_lamports();
+        let msol_supply = self.state.msol_supply;
+
+        check_token_source_account(
+            &self.burn_msol_from,
+            self.burn_msol_authority.key,
+            msol_amount,
+        )?;
+
+        let mut stake = self.state.stake_system.get_checked(
+            &self.stake_list.data.as_ref().borrow(),
+            stake_index,
+            self.stake_account.to_account_info().key,
+        )?;
+        let last_update_stake_delegation = stake.last_update_delegated_lamports;
+
+        // require the stake is not in emergency_unstake
+        require_neq!(
+            stake.is_emergency_unstaking,
+            0,
+            MarinadeError::StakeAccountIsEmergencyUnstaking
+        );
+
+        // require stake is active (deactivation_epoch == u64::MAX)
+        let delegation = self.stake_account.delegation().ok_or_else(|| {
+            error!(MarinadeError::RequiredDelegatedStake).with_account_name("stake_account")
+        })?;
+        require_eq!(
+            delegation.deactivation_epoch,
+            std::u64::MAX,
+            MarinadeError::RequiredActiveStake
+        );
+
+        let mut validator = self
+            .state
+            .validator_system
+            .get(&self.validator_list.data.as_ref().borrow(), validator_index)?;
+
+        // check currently_staked in this account & validator vote-key
+        check_stake_amount_and_validator(
+            &self.stake_account,
+            stake.last_update_delegated_lamports,
+            &validator.validator_account,
+        )?;
+
+        // compute how many lamport the burned mSOL represents
+        require_gt!(msol_amount, self.state.min_withdraw);
+        let sol_value = self.state.calc_lamports_from_msol_amount(msol_amount)?;
+        // apply withdraw_stake_account_fee to avoid economical attacks
+        // withdraw_stake_account_fee must be >= one epoch staking rewards
+        let withdraw_stake_account_fee_lamports =
+            self.state.withdraw_stake_account_fee.apply(sol_value);
+        // the fee value will be burned but not delivered, thus increasing mSOL value slightly for all mSOL holders
+        let split_lamports = sol_value - withdraw_stake_account_fee_lamports;
+        // check withdraw amount >= WithdrawAmountIsTooLow
+        require_gte!(
+            split_lamports,
+            self.state.min_withdraw,
+            MarinadeError::WithdrawAmountIsTooLow
+        );
+        const MIN_STAKE_ACCOUNT_LAMPORTS: u64 = LAMPORTS_PER_SOL;
```

Comment:
We have state.stake_system.min_stake for this. And do not use 1 SOL because it will not work in solana 1.16 (need to be > 1 SOL)